### DISCUSSION TITLE
iOS编译问题未生成fat .a文件

### DISCUSSION THREAD

**Discussion by bluedaquiri** [alibaba/xquic#45] [Q&A] [answered]
@RuiqiZhou 

编译步骤：
1.clone xquic源码
 >  git clone git@github.com:alibaba/xquic.git
> cd xquic

2.clone boringssl源码
> git clone git@github.com:BabaSSL/BabaSSL.git ./third_party/babassl

3.执行脚本
> sh xqc_build.sh ios build-ios build-ios-target ./third_party/boringssl

> 脚本修改 
1.删除-DBORINGSSL_PREFIX=bs
          -DBORINGSSL_PREFIX_SYMBOLS=$cur_dir/bssl_symbols.txt
2.修改-DENABLE_BITCODE=1

问题描述：
文件路径: 
> ios/xquic/xquic/Libs 未生成.a文件
> build-ios-target中所有架构的.a通过lipo -info查看都是libxquic.a is architecture: x86_64

脚本执行过程报错:
`cp: /Users/bluedaquiri/Desktop/Quic2/xquic/build-ios/outputs/*.so: No such file or directory
fatal error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: specifed architecture type (armv7) for file (/Users/bluedaquiri/Desktop/Quic2/xquic/build-ios-target/armv7/libxquic.a) does not match its cputype (16777223) and cpusubtype (3) (should be cputype (12) and cpusubtype (9))
fatal error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: specifed architecture type (armv7) for file (/Users/bluedaquiri/Desktop/Quic2/xquic/build-ios-target/armv7/libcrypto.a) does not match its cputype (16777223) and cpusubtype (3) (should be cputype (12) and cpusubtype (9))
fatal error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: specifed architecture type (armv7) for file (/Users/bluedaquiri/Desktop/Quic2/xquic/build-ios-target/armv7/libssl.a) does not match its cputype (16777223) and cpusubtype (3) (should be cputype (12) and cpusubtype (9))
bluedaquiri@cmw-118 xquic % lipo -info /Users/bluedaquiri/Desktop/Quic2/xquic/build-ios/outputs/libxquic.a 
Non-fat file: /Users/bluedaquiri/Desktop/Quic2/xquic/build-ios/outputs/libxquic.a is architecture: x86_64`

> 修改后的脚本
```
# Copyright (c) 2022, Alibaba Group Holding Limited
#!/bin/sh

android_archs=(armeabi-v7a arm64-v8a)
ios_archs=(armv7 arm64 x86_64)
cur_dir=$(cd "$(dirname "$0")";pwd)

cp -f $cur_dir/cmake/CMakeLists.txt  $cur_dir/CMakeLists.txt

platform=$1
build_dir=$2
artifact_dir=$3

# boringssl is used as default
ssl_type="boringssl"
ssl_path=$4
ssl_lib_path="${ssl_path}/"

if [ -z "$ssl_path" ] || [ -z "$ssl_lib_path" ] ; then
    echo "ssl environment not specified"
    exit 0
fi

create_dir_or_exit() {
    if [ x"$2" == x ] ; then
        echo "$1 MUST NOT be empty"
        exit 1
    fi
    if [ -d $2 ] ; then
        echo "directory already exists"
    else
        mkdir $2
        echo "create $1 directory($2) suc"
    fi
}

platform=$(echo $platform | tr A-Z a-z )

if [ x"$platform" == xios ] ; then 
    if [ x"$IOS_CMAKE_TOOLCHAIN" == x/cmake/ios.toolchain.cmake ] ; then
        echo "IOS_CMAKE_TOOLCHAIN MUST be defined"
        exit 0
    fi

    archs=${ios_archs[@]} 
    configures="-DSSL_TYPE=${ssl_type}
                -DSSL_PATH=${ssl_path}
                -DSSL_LIB_PATH=${ssl_lib_path}
                -DDEPLOYMENT_TARGET=10.0
                -DCMAKE_BUILD_TYPE=Minsizerel
                -DXQC_ENABLE_TESTING=OFF
                -DXQC_BUILD_SAMPLE=OFF
                -DGCOV=OFF
                -DCMAKE_TOOLCHAIN_FILE=${IOS_CMAKE_TOOLCHAIN}
                -DENABLE_BITCODE=1
                -DXQC_NO_SHARED=1"

elif [ x"$platform" == xandroid ] ; then
    if [ x"$ANDROID_NDK" == x ] ; then
        echo "ANDROID_NDK MUST be defined"
        exit 0
    fi

    archs=${android_archs[@]}
    configures="-DSSL_TYPE=${ssl_type}
                -DSSL_PATH=${ssl_path}
                -DSSL_LIB_PATH=${ssl_lib_path}
                -DCMAKE_BUILD_TYPE=Minsizerel
                -DXQC_ENABLE_TESTING=OFF
                -DXQC_BUILD_SAMPLE=OFF
                -DGCOV=OFF
                -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake
                -DANDROID_STL=c++_shared
                -DANDROID_NATIVE_API_LEVEL=android-19
                -DXQC_DISABLE_RENO=OFF
                -DXQC_ENABLE_BBR2=ON
                -DXQC_DISABLE_LOG=ON"
else 
    echo "no support platform"
    exit 0
fi


generate_plat_spec() {
    plat_spec=
    if [ x"$platform" == xios ] ; then
        plat_spec="-DARCHS=$1"
        if [ x"$1" == xx86_64 ] ; then
            plat_spec="$plat_spec -DPLATFORM=SIMULATOR64"
        elif [ x"$1" == xi386 ] ; then
            plat_spec="$plat_spec -DPLATFORM=SIMULATOR"
        fi
    else
        plat_spec="-DANDROID_ABI=$1"
    fi
    echo $plat_spec
}

create_dir_or_exit build $build_dir
# to absoulute path 
build_dir=$cur_dir/$build_dir

create_dir_or_exit artifact $artifact_dir
artifact_dir=$cur_dir/$artifact_dir

cd $build_dir 

for i in ${archs[@]} ;
do
    rm -f  CMakeCache.txt
    rm -rf CMakeFiles
    rm -rf Makefile
    rm -rf cmake_install.cmake
    rm -rf include
    rm -rf outputs
    rm -rf third_party

    echo "compiling xquic on $i arch"
    cmake  $configures  $(generate_plat_spec $i ) -DLIBRARY_OUTPUT_PATH=`pwd`/outputs/ ..
    make -j 4
    if [ $? != 0 ] ; then
        exit 0
    fi

    if [ ! -d  ${artifact_dir}/$i ] ; then
        mkdir -p ${artifact_dir}/$i
    fi
    cp -f `pwd`/outputs/*.a     ${artifact_dir}/$i/
    cp -f `pwd`/outputs/*.so    ${artifact_dir}/$i/
done


make_fat() {
    script="lipo -create"
    for i in ${archs[@]} ;
    do
        script="$script -arch $i $artifact_dir/$i/$1  "
    done
    script="$script -output $cur_dir/ios/xquic/xquic/Libs/$1"
    $($script) 
}


if [ x"$platform" == xios ] ; then
    if [ ! -d $cur_dir/ios/xquic/xquic/Headers ] ; then
        mkdir -p $cur_dir/ios/xquic/xquic/Headers
    fi
    if [ ! -d $cur_dir/ios/xquic/xquic/Libs ] ; then
        mkdir -p $cur_dir/ios/xquic/xquic/Libs
    fi
    make_fat libxquic.a
    make_fat libcrypto.a
    make_fat libssl.a
    cp -f $cur_dir/include/xquic/*   $cur_dir/ios/xquic/xquic/Headers/
    cp -f $build_dir/include/xquic/* $cur_dir/ios/xquic/xquic/Headers/

fi
```

**Comment by ruiqizhou** [comment]
@lurker-Chen

**Comment by lurker-Chen** [comment, marked as answer]
Are there any .a files of other architectures in /build-ios-target?

**Reply by bluedaquiri** [reply]
感谢回复
是的，所有文件夹中都存在.a，但是都是x86的

<img width="493" alt="截屏2022-01-12 18 45 08" src="https://user-images.githubusercontent.com/17403572/149126030-01a6e53e-1caa-4d20-81ec-8ab33d09c244.png">

`lipo -info /Users/bluedaquiri/Desktop/Quic2/xquic/build-ios-target/armv7/libxquic.a
Non-fat file: /Users/bluedaquiri/Desktop/Quic2/xquic/build-ios-target/armv7/libxquic.a is architecture: x86_64`

此外我还想支持Mac M1的模拟器，支持arm64架构的模拟器，我该怎么修改sh文件？

**Reply by lurker-Chen** [reply]
can you delete x86_64 in ios_archs (xquic_build.sh:5) and try again?

**Reply by lurker-Chen** [reply]
if you want to support other architectures, change ios_archs and generate_plat_spec in xquic_build.sh
![image](https://user-images.githubusercontent.com/97525334/149132476-aded8639-f9f3-46d7-9ea8-62ab3960ca76.png)

**Reply by bluedaquiri** [reply]
感谢回复

模拟器x86架构和真机的arm7和arm64架构都可以成功编译

模拟器的arm64架构编译遇到了些问题，不管是OS64COMBINED还是SIMULATOR64，麻烦帮忙看一下

脚本修改:
```
ios_archs=(arm64 os64combined)

generate_plat_spec() {
    plat_spec=
    if [ x"$platform" == xios ] ; then
        plat_spec="-DARCHS=$1"
        if [ x"$1" == xarm64 ] ; then
            plat_spec="$plat_spec -DPLATFORM=OS64"
        elif [ x"$1" == xos64combined ] ; then
            plat_spec="$plat_spec -DPLATFORM=OS64COMBINED"
        fi
    else
        plat_spec="-DANDROID_ABI=$1"
    fi
    echo $plat_spec
}

```

执行报错:

```
CMake Error at cmake/ios.toolchain.cmake:321 (message):
  The COMBINED options only work with Xcode generator, -G Xcode
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.21.1/share/cmake/Modules/CMakeDetermineSystem.cmake:124 (include)
  CMakeLists.txt:4 (project)
```

**Reply by lurker-Chen** [reply]
compile in OS64COMBINED should add "-G Xcode" to configures(xqc_build.sh:46)
if you want to support other architectures, you should check ios.toolchain.cmake.
![image](https://user-images.githubusercontent.com/97525334/149701154-a32863bb-d7a8-4bca-adb2-37b300d42611.png)

**Reply by xilin** [reply]
Update to newer version of ios-cmake, then update `ios_archs` and `generate_plat_spec`. Note that you should create xcframework instead of fat .a file nowadays

```
ios_archs=(sim_arm64 arm64 x86_64)
...
generate_plat_spec() {
    if [ x"$platform" == xios ] ; then
        plat_spec="-DARCHS=$1"
        if [ x"$1" == xx86_64 ] ; then
            plat_spec="$plat_spec -DPLATFORM=SIMULATOR64"
        elif [ x"$1" == xarm64 ] ; then
            plat_spec="$plat_spec -DPLATFORM=OS64"
        elif [ x"$1" == xsim_arm64 ] ; then
            plat_spec="-DARCHS=arm64 -DPLATFORM=SIMULATORARM64"
        fi
    elif [ x"$platform" == xharmony ] ; then
        plat_spec="-DOHOS_ARCH=$1"
    else
        plat_spec="-DANDROID_ABI=$1"
    fi
    echo $plat_spec
}
```
